### PR TITLE
Dynamically import polyfills for Safari

### PR DESCRIPTION
**Pull request by swissspidy**
## Context

The editor bundles some polyfills to improve Safari support.

Unfortunately these are part of the main bundle, even for browsers that don't really need the polyfills.

These should be split up and loaded conditionally.

## Summary

Still need to do some testing, but making these dynamic imports (i.e. lazy load) should prevent loading this polyfill on Chrome and other browsers. 🚀

`ResizeObserver` pollyfill:

This way it will basically only be loaded on Safari ≤ 13, which is technically not supported by the project. However, given how easy it is to maintain support in accordance with the project’s browser support policies, this can be kept for now.

`:focus-visible` polyfill:

Since this is basically only needed for Safari, lazy loading this optimizes things for all other browsers.

## To-do

- [x] Figure out why `focus-visible` chunk is missing from `edit-story.chunks.php` (causes 404s)
- [x] Figure out why the `resize-observer-polyfill` chunk is still loadeded in browsers that don't actually need it 

### QA

- [x] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

### UAT

- [x] UAT should use the same steps as above.

This PR can be tested by following these steps:

1.

## Checklist

- [ ] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

See #8251

### CODE REVIEW

**Comment by miina** [comment]
@swissspidy Looks like the tests are failing.

File: assets/src/design-system/theme/global/overrideFocusOutline.js [hunk 1]
```diff
@@ -20,6 +20,10 @@
 import { createGlobalStyle, css } from 'styled-components';
 import 'focus-visible';
```
Comment (miina) [inline assets/src/design-system/theme/global/overrideFocusOutline.js]: Is this line still necessary as well?

File: assets/src/design-system/theme/global/overrideFocusOutline.js [hunk 2]
```diff
@@ -20,6 +20,10 @@
...
 
+import(/* webpackChunkName: "focus-visible" */ 'focus-visible').catch(
```
Comment (miina) [inline assets/src/design-system/theme/global/overrideFocusOutline.js]: Saw you mentioned lazy loading this in the description, should it also have `webpackMode: "lazy"` then?

File: assets/src/design-system/theme/global/overrideFocusOutline.js (hunk 1 above)
Comment (swissspidy) [reply assets/src/design-system/theme/global/overrideFocusOutline.js]: Nope! Forgot to remove 😅

File: assets/src/design-system/theme/global/overrideFocusOutline.js (hunk 2 above)
Comment (swissspidy) [reply assets/src/design-system/theme/global/overrideFocusOutline.js]: That's the [default](https://webpack.js.org/api/module-methods/#magic-comments)

**Review by miina — APPROVED** [review]

**Review by BrittanyIRL — COMMENTED** [review]
When testing for the desired `focus-visible` styles in chrome (my version: `90.0.4430.212`) on this branch, the blue outline we want is actually now completely gone?

https://user-images.githubusercontent.com/10720454/120223654-c8358b80-c1f6-11eb-9980-aa3255c7068b.mp4

Looks good in safari though! My default accessibility color palette takes over but it is consistent.

**Comment by swissspidy** [comment]
@BrittanyIRL Yeah unfortunately I found out later that the webpack configuration needs some updating, as there's a 404 trying to load the `focus-visible` polyfill. I've noted that in the to-do above & made the PR a draft again.

While webpack seems to create a separate JS file for this `focus-visible` chunk, it is not referenced in the `*.chunks.php` file, which is quite strange. I thought it might be due to the polyfill not meeting the size thresholds, but changing that didn't work.

If you are an experienced webpack wrangler or know someone on the team who is, any feedback is appreciated!

**Comment by BrittanyIRL** [comment]
> @BrittanyIRL Yeah unfortunately I found out later that the webpack configuration needs some updating, as there's a 404 trying to load the `focus-visible` polyfill. I've noted that in the to-do above & made the PR a draft again.
> 
> While webpack seems to create a separate JS file for this `focus-visible` chunk, it is not referenced in the `*.chunks.php` file, which is quite strange. I thought it might be due to the polyfill not meeting the size thresholds, but changing that didn't work.
> 
> If you are an experienced webpack wrangler or know someone on the team who is, any feedback is appreciated!

oh yup you did note that, sorry about that! 

I'll shop the issue around for some extra eyes, see if anyone on our end has suggestions.

**Comment by spacedmonkey** [comment]
@swissspidy Is this going to be worked as part of performance work?

**Comment by swissspidy** [comment]
@spacedmonkey If you want to take a stab at fixing the webpack config, feel free.
Just timebox it to a couple of hours or so, as I don't think it's critical for the current effort.

**Comment by swissspidy** [comment]
Oh I think I found the problem!

**Comment by swissspidy** [comment]
> `assets/js/chunk-resize-observer-polyfill-********************.js`	2.55 kB	+2.55 kB (new file)	🆕
> `assets/js/vendors-chunk-resize-observer-polyfill-********************.js`	2.55 kB	+2.55 kB (new file)	🆕

Hmm why are there two files now 🤔

**Comment by swissspidy** [comment]
> > `assets/js/chunk-resize-observer-polyfill-********************.js`	2.55 kB	+2.55 kB (new file)	🆕
> > `assets/js/vendors-chunk-resize-observer-polyfill-********************.js`	2.55 kB	+2.55 kB (new file)	🆕
> 
> Hmm why are there two files now 🤔

I suppose one is coming from `edit-story` and one is coming from the design system (and thus `node_modules`), so it lands in `vendors`

This could be prevented if `ResizeObserver` would only be used in design system and not edit-story, or vice-versa, or if we don't load polyfills in design system but only edit-story.


1. `syncOuterHeightWithInner` could be replaced with `useResizeEffect`
2. For `TextPane`, it is not clear to me at all why `ResizeObserver` is used and what it does there...

**Review by BrittanyIRL — APPROVED** [review]
🎉 🎉  You fixed it! 🎉 🎉  

Works great. Tested in Chrome and Safari.

**Comment by swissspidy** [comment]
> > > `assets/js/chunk-resize-observer-polyfill-********************.js`	2.55 kB	+2.55 kB (new file)	🆕
> > > `assets/js/vendors-chunk-resize-observer-polyfill-********************.js`	2.55 kB	+2.55 kB (new file)	🆕
> > 
> > 
> > Hmm why are there two files now 🤔
> 
> I suppose one is coming from `edit-story` and one is coming from the design system (and thus `node_modules`), so it lands in `vendors`
> 
> This could be prevented if `ResizeObserver` would only be used in design system and not edit-story, or vice-versa, or if we don't load polyfills in design system but only edit-story.
> 
> 1. `syncOuterHeightWithInner` could be replaced with `useResizeEffect`
> 2. For `TextPane`, it is not clear to me at all why `ResizeObserver` is used and what it does there...

Rewriting these worked like a charm.

Turns out the second chunk was coming from `web-stories-block`, which was importing that via design system / dashboard.

It should get away again after #8339 is merged.

**Pull request merged**